### PR TITLE
Create service which can get list of links from one page with/without JS

### PR DESCRIPTION
**Pull request by aleksasha-o**

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -1 +1,6 @@
+# frozen_string_literal: true
 
+require 'nokogiri'
+require 'open-uri'
+require 'ferrum'
+require_relative 'services/products_links/get_products_links'
```
Comment (olegsobchuk) [inline main.rb]: required/included files should be in use somehow. So add something here, what runs some code and provides some result

Comment (olegsobchuk) [reply main.rb]: so I don't think that 
```
require 'nokogiri'
require 'open-uri'
require 'ferrum'
```
are needed here

File: services/products_links/get_products_links.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class GetProductsLinks
+  RETRY_INTERVAL = 0.01
+  TIMEOUT = 5
+
+  class NotFoundError < StandardError; end
+
+  def initialize(url:, tag:)
+    @url = url
+    @tag = tag
+    @browser = Ferrum::Browser.new
```
Comment (olegsobchuk) [inline services/products_links/get_products_links.rb]: this should be in separate method, not in parameters

File: services/products_links/get_products_links.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def parse_links_without_js
+    doc = Nokogiri::HTML(URI.open(@url))
```
Comment (olegsobchuk) [inline services/products_links/get_products_links.rb]: why we should use `Nokogiri` here? Is it possible to use `Ferrum` everywhere?

File: services/products_links/get_products_links.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+class GetProductsLinks
+  RETRY_INTERVAL = 0.01
```
Comment (olegsobchuk) [inline services/products_links/get_products_links.rb]: pls, decrease it to `0.3`

**Review by olegsobchuk — COMMENTED** [review]
Review 1

File: services/products_links/get_products_links.rb (hunk 3 above)
Comment (aleksasha-o) [reply services/products_links/get_products_links.rb]: Nokogiri is more efficient and faster. It just goes to the page and gives the html, while Ferrum brings up a whole invisible browser window to read from it. But, unfortunately, Nokogiri doesn't know how to work with JS, that why we need Ferrum

File: services/link_parsers/get_page.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+require 'ferrum'
+
+class GetPage
```
Comment (olegsobchuk) [inline services/link_parsers/get_page.rb]: if you have scopes for naming like `link_parsers/get_page.rb` so use 
```
module LinkParsers
  class GetPage
```

File: services/browser.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      raise NotFoundError unless @browser.at_css(@tag)
+    rescue NotFoundError
+      sleep RETRY_INTERVAL
+      load_time += RETRY_INTERVAL
+      retry if load_time <= TIMEOUT
+      raise "Couldn't find element #{@tag} on #{@url}"
+    end
+  end
+
+  def save_body
```
Comment (olegsobchuk) [inline services/browser.rb]: rename to `body`

File: services/browser.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  private
+
+  def browser
+    @browser ||= Ferrum::Browser.new(timeout: 15)
+  end
+
+  def wait_for_element
+    load_time = 0
+    begin
```
Comment (olegsobchuk) [inline services/browser.rb:37]: use `loop` here2
```
loop do
  exit if load_time > ....
  
  # do something
end
```

File: services/browser.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      raise NotFoundError unless @browser.at_css(@tag)
+    rescue NotFoundError
+      sleep RETRY_INTERVAL
+      load_time += RETRY_INTERVAL
+      retry if load_time <= TIMEOUT
+      raise "Couldn't find element #{@tag} on #{@url}"
```
Comment (olegsobchuk) [inline services/browser.rb]: don't raise error here

File: services/models/shop_hodinkee_model.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+module Models
+  class ShopHodinkeeModel < BaseModel
+    attr_accessor :water_resistance
+
+    def initialize(**args)
+      super
+      @water_resistance = water_resistance
+    end
+
```
Comment (olegsobchuk) [inline services/models/shop_hodinkee_model.rb]:
```
HOST = 'https://qwewqeqweqw.com'
PAGINATION_SELECTOR = '.paginat'
```

File: services/processors/shop_hodinkee.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+require_relative '../models/shop_hodinkee_model'
+
+module Processors
+  class ShopHodinkee
+    def initialize(page: 1)
+      @page = page
+      @links = []
+    end
+    
+    def call
```
Comment (olegsobchuk) [inline services/processors/shop_hodinkee.rb]:
```
content = browser.visit(url)
items_collection = parser(content).items_collection
items_collection.each do |item_page|
  item_content = browser.visit(item_page)
  attrs = parser.item(item_content)
  item = model.new(attrs)
  # imagine we save it to somewhere
  # item.save
end
browser.... # visit next page
```

File: services/processors/shop_hodinkee.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      collect_links
+      collect_data
+    end
+
+    private
+
```
Comment (olegsobchuk) [inline services/processors/shop_hodinkee.rb]: def browser
  @browser ||= Browser.new(target)
end

File: services/processors/shop_hodinkee.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    def collect_links
+      page = "https://shop.hodinkee.com/collections/watches?page=#{@page}"
+      puts "Parsing #{page}"
+      opened_page = Browser.new(url: page, tag: '.product-title').visit
+      parser = Parser.new(opened_page)
+      parser.parse_links('.product-title').each do |item|
```
Comment (olegsobchuk) [inline services/processors/shop_hodinkee.rb]: parser.product_links

File: services/models/shop_hodinkee_model.rb (hunk 9 above)
Comment (olegsobchuk) [reply services/models/shop_hodinkee_model.rb]: move it to Processor, pls

File: services/processors/shop_hodinkee.rb [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    
+    def call
+      collect_links
+      collect_data
+    end
+
+    private
+
+    def collect_links
+      page = "https://shop.hodinkee.com/collections/watches?page=#{@page}"
```
Comment (olegsobchuk) [inline services/processors/shop_hodinkee.rb]: use URI builder

File: main.rb [hunk 14]
```diff
@@ -3,10 +3,4 @@
 require_relative 'services/parser'
 require_relative 'services/processors/shop_hodinkee'
 
-# @example = Browser.new(url: 'https://www.bobswatches.com/shop', tag: '.item form a').visit
-# puts Parser.new(url: @example, tag: '.item form a')
-#                         .parse_links
-#                         .map { |item| "https://www.bobswatches.com/shop#{item}" }
-
-puts "_______________________________________________________________________________"
 puts Processors::ShopHodinkee.new.call
```
Comment (olegsobchuk) [inline main.rb]: `.new.call` -> `.call`

```
def self.call
  new.call
end
```

File: services/browser.rb [hunk 15]
```diff
@@ -7,16 +7,20 @@ class Browser
 
   class NotFoundError < StandardError; end
 
-  def initialize(url:, tag:)
-    @url = url
+  def visit(url:, tag:)
     @tag = tag
-  end
+    @load_time = 0
```
Comment (olegsobchuk) [inline services/browser.rb]: use local vars instead of instance

File: services/browser.rb [hunk 16]
```diff
@@ -7,16 +7,20 @@ class Browser
...
+
+    browser.network.intercept
+    browser.on(:request) do |request|
+      next request.abort if request.match?(/\.png|\.jpg|\.jpeg|\.svg|\.woff2/)
```
Comment (olegsobchuk) [inline services/browser.rb]: use `/\.png|\.jpg|\.jpeg|\.svg|\.woff2/` as CONSTANT

File: services/browser.rb (hunk 15 above)
Comment (olegsobchuk) [inline services/browser.rb]: use `attr_accessor` for this

File: services/browser.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+      request.continue
+    end
+    browser.go_to(url)
+    wait_for_element
+
+    body
+    browser.quit
```
Comment (olegsobchuk) [inline services/browser.rb:24]: what if body rise error? Will `broser.quit` work?

File: services/browser.rb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+
+  def browser
+    @browser ||= Ferrum::Browser.new(pending_connection_errors: false)
+  end
+
+  def wait_for_element
+    raise NotFoundError if @browser.at_css(@tag).nil?
+  rescue NotFoundError, Ferrum::NodeNotFoundError
+    sleep RETRY_INTERVAL
+    @load_time += RETRY_INTERVAL
```
Comment (olegsobchuk) [inline services/browser.rb]: use `attr_accessor`

File: services/browser.rb [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+    retry if @load_time <= TIMEOUT
+  end
+
+  def body
+    @body = browser.body
```
Comment (olegsobchuk) [inline services/browser.rb:47]: attr_accessor

File: services/models/base_model.rb [hunk 20]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Models
+  class BaseModel
+    attr_accessor :brand, :model, :price, :sku, :dial_color,
```
Comment (olegsobchuk) [inline services/models/base_model.rb]: use list of available base attributes as global var
```
ATTRIBUTES = [:one, :two, ....]
```
then use this list in your further code

File: services/models/shop_hodinkee_model.rb [hunk 21]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+require_relative 'base_model'
+
+module Models
+  class ShopHodinkeeModel < BaseModel
+    attr_accessor :water_resistance
+
+    def initialize(**args)
+      super
+      @water_resistance = water_resistance
```
Comment (olegsobchuk) [inline services/models/shop_hodinkee_model.rb]: !

**Review by olegsobchuk — APPROVED** [review]
LGTM

**Pull request merged**